cpp/src/parquet/arrow/reader_internal.cc

```diff
@@ -685,21 +687,148 @@ Status BuildSchemaManifest(const SchemaDescriptor* schema,
+  *max = ::arrow::MakeScalar(static_cast<CType>(typed_statistics->max()));
+  return Status::OK();
+}
+
+template <typename StatisticsType>
+Status MakeMinMaxIntegralScalar(const std::shared_ptr<Statistics>& statistics,
+                                std::shared_ptr<::arrow::Scalar>* min,
+                                std::shared_ptr<::arrow::Scalar>* max) {
+  const auto column_desc = statistics->descr();
+  const auto& logical_type = column_desc->logical_type();
+  const auto& integer = checked_pointer_cast<const IntLogicalType>(logical_type);
+  const bool is_signed = integer->is_signed();
+
+  switch (integer->bit_width()) {
+    case 8:
+      return is_signed ? MakeMinMaxScalar<int8_t, StatisticsType>(statistics, min, max)
+                       : MakeMinMaxScalar<uint8_t, StatisticsType>(statistics, min, max);
+    case 16:
+      return is_signed ? MakeMinMaxScalar<int16_t, StatisticsType>(statistics, min, max)
+                       : MakeMinMaxScalar<uint16_t, StatisticsType>(statistics, min, max);
+    case 32:
+      return is_signed ? MakeMinMaxScalar<int32_t, StatisticsType>(statistics, min, max)
+                       : MakeMinMaxScalar<uint32_t, StatisticsType>(statistics, min, max);
+    case 64:
+      return is_signed ? MakeMinMaxScalar<int64_t, StatisticsType>(statistics, min, max)
+                       : MakeMinMaxScalar<uint64_t, StatisticsType>(statistics, min, max);
+  }
+
+  return Status::OK();
+}
+
+template <typename StatisticsType>
+Status FromParquetStatisticsConverted(const std::shared_ptr<Statistics>& statistics,
+                                      std::shared_ptr<::arrow::Scalar>* min,
+                                      std::shared_ptr<::arrow::Scalar>* max) {
+  return Status::NotImplemented("Extract statistics to arrow::Scalar not implemented");
+}
+
+template <>
+Status FromParquetStatisticsConverted<Int32Statistics>(
+    const std::shared_ptr<Statistics>& statistics, std::shared_ptr<::arrow::Scalar>* min,
+    std::shared_ptr<::arrow::Scalar>* max) {
+  auto column_desc = statistics->descr();
+  auto logical_type = column_desc->logical_type();
+
+  switch (logical_type->type()) {
+    case LogicalType::Type::INT:
+      return MakeMinMaxIntegralScalar<Int32Statistics>(statistics, min, max);
+    default:
+      return Status::NotImplemented("Cannot extract from Int32Statistics");
+  }
+
+  return Status::OK();
+}
+
+template <>
+Status FromParquetStatisticsConverted<Int64Statistics>(
+    const std::shared_ptr<Statistics>& statistics, std::shared_ptr<::arrow::Scalar>* min,
+    std::shared_ptr<::arrow::Scalar>* max) {
+  auto column_desc = statistics->descr();
+  auto logical_type = column_desc->logical_type();
+
+  switch (logical_type->type()) {
+    case LogicalType::Type::INT:
+      return MakeMinMaxIntegralScalar<Int64Statistics>(statistics, min, max);
+    default:
+      return Status::NotImplemented("Cannot extract from Int64Statistics");
+  }
```

Comment:
Since `FromParquetStatisticsConverted` is an internal function and the two non-failure overloads are mostly identical, please fold them into a single generic definition